d3b_release_maker/release_maker.py

```diff
@@ -0,0 +1,407 @@
+                    )
+            else:
+                break
+
+        return response
+
+    def yield_paginated(self, endpoint, query_params):
+        """
+        Yield from paginated endpoint
+        """
+        query_params.update({"page": 1, "per_page": 100})
+        items = True
+        while items:
+            items = self.get(endpoint, params=query_params).json()
+            yield from items
+            query_params["page"] += 1
+
+
+class GitHubReleaseNotes:
+    def __init__(self):
+        self.session = GitHubSession()
+
+    def _starting_emojis(self, title):
+        """
+        Detect emojis at the start of a PR title (and fix malformed titles)
+        """
+        emojis = set()
+        graphemes = regex.findall(r"\X", title)
+        for i, g in enumerate(graphemes):
+            if any(char in emoji.UNICODE_EMOJI for char in g):
+                emojis.add(g)
+            else:  # stop after first non-hit
+                if g != " ":
+                    # fix missing space in malformed titles
+                    title = (
+                        "".join(graphemes[:i]) + " " + "".join(graphemes[i:])
+                    )
+                break
+
+        return (emojis or {"?"}, title)
+
+    def _get_merged_prs(self, after):
+        """
+        Get all non-release PRs merged into master after the given time
+        """
+        print(f"Fetching PRs ...")
+        endpoint = f"{self.base_url}/pulls"
+        query_params = {"base": "master", "state": "closed"}
+        prs = []
+        for p in self.session.yield_paginated(endpoint, query_params):
+            if p["merged_at"]:
+                if p["merged_at"] < after:
+                    break
+                elif regex.search(release_pattern, p["title"]) is None:
+                    prs.append(p)
+        return prs
+
+    def _get_commit_date(self, commit_url):
+        """
+        Get date of commit at commit_url
+        """
+        commit = self.session.get(commit_url).json()
+        return commit["commit"]["committer"]["date"]
+
+    def _get_last_tag(self):
+        """
+        Get latest tag and when it was committed
+        """
+        tags = self.session.get(f"{self.base_url}/tags").json()
+
+        # Get latest commit of last tagged release
+        if len(tags) > 0:
+            for t in tags:
+                try:
+                    prefix, version = split_at_pattern(t["name"], r"\d")
+                    # Raise on non-semver tags so we can skip them
+                    semver.VersionInfo.parse(version)
+                    return {
+                        "name": t["name"],
+                        "date": self._get_commit_date(t["commit"]["url"]),
+                        "commit_sha": t["commit"]["sha"],
+                    }
+                except ValueError:
+                    pass
+        else:
+            return None
+
+    def _next_release_version(self, prev_version, release_type):
```

Comment:
May be able to discard this in favor of just using `semver`'s `bump_major()` style functions.

Reply:
Actually I'm pretty sure I _was_ using semver's functions in this code at some point. I wonder when that got flipped.

Reply:
73355d56ad5cf2a6150af99031a8ad64b1f932e5